### Part I: Pricing Analysis
### Optimal pricing for weekend vs. weekday

1) Estimate the demand d(p) by regressing on price and weekend. Interpret the regression results. Provide an interpretation in your own words!

In [16]:
import pandas as pd
df_demand = pd.read_csv("demand_data.csv")
df_demand
df_demand.groupby('weekend').demand.mean()

weekend
0    5.913043
1    6.888889
Name: demand, dtype: float64

In [17]:
import statsmodels.formula.api as smf
import numpy as np
weekend_fit = smf.ols(formula='demand ~ price + (weekend == 1)', data=df_demand).fit()
weekend_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 demand   R-squared:                       0.881
Model:                            OLS   Adj. R-squared:                  0.873
Method:                 Least Squares   F-statistic:                     107.1
Date:                Sun, 30 Apr 2023   Prob (F-statistic):           4.07e-14
Time:                        20:11:34   Log-Likelihood:                -27.363
No. Observations:                  32   AIC:                             60.73
Df Residuals:                      29   BIC:                             65.12
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept                9.9293      0.312     31.808      0.000       9.291      10.568
weekend == 1[T.True]     0.4198      0.238      1.762      0.089      -0.068       0.907
price                   -0.8772      0.063    -14.033      0.000      -1.005      -0.749
==============================================================================
Omnibus:                        4.720   Durbin-Watson:                   2.107
Prob(Omnibus):                  0.094   Jarque-Bera (JB):                3.505
Skew:                           0.797   Prob(JB):                        0.173
Kurtosis:                       3.297   Cond. No.                         15.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Answer: According to regression table shows above, the coefficient of price is negative which means this is make sense. Because everytime we increase the price for one unit, the demand will go down for 0.8772 units. The coefficient of weekend is 0.4198 which means if we record demand on weekend, the relationship between demand and if this recording on weekend.
    

Calculate the optimal weekday price p* (when w = 1) and its corresponding profit 

Clearly state your findings, do not simply show output of running code

Determine whether the optimal weekday price is within the credible range and
provide response in your own words!

In [18]:
import pandas as pd
df_demand = pd.read_csv("demand_data.csv")
df_demand
df_demand.groupby('weekend').demand.mean()

weekend
0    5.913043
1    6.888889
Name: demand, dtype: float64

In [40]:
weekend_fit.params

Intercept               9.929305
weekend == 1[T.True]    0.419832
price                  -0.877246
dtype: float64

In [20]:
pi = lambda x: -(weekend_fit.params[0] + weekend_fit.params[1]*1 + weekend_fit.params[2]*x ) * (x-4)

In [21]:
from scipy.optimize import minimize_scalar
#optimal_price = 
minimize_scalar(pi, method='brent')
#print(optimal_price)


     fun: -13.33368361942117
    nfev: 10
     nit: 4
 success: True
       x: 7.898650699949416

In [41]:
df_demand.price.min()

1.5

In [42]:
df_demand.price.max()

7.5

#### Answer: the optimal weekend price p is 7.89 when w = 1. The optimal weekday price is not within the credible range, since our range is between 1.5 and 7.5. Price(12.85 > max range 7.5). The optimal profit is 13.333 dollars.

Calculate the optimal weekend price p* (when w = 0) and its corresponding profit

o Clearly state your findings, do not simply show output of running code

o Determine whether the optimal weekend price is within the credible range and
provide response in your own words!

In [24]:
#demand_fit = smf.ols(formula='demand ~ price + (weekend == 1)', data=df_demand).fit()
#demand_fit.summary()
#demand_fit.params

Intercept               9.929305
weekend == 1[T.True]    0.419832
price                  -0.877246
dtype: float64

In [43]:
pi2 = lambda x: -(weekend_fit.params[0] + weekend_fit.params[1]*0 + weekend_fit.params[2]*x ) * (x-4)

In [44]:
from scipy.optimize import minimize_scalar
#optimal_price is x
minimize_scalar(pi2, method='brent')

     fun: -11.747134712211084
    nfev: 10
     nit: 4
 success: True
       x: 7.659360769101051

In [45]:
df_demand.price.min()

1.5

In [46]:
df_demand.price.max()

7.5

#### Answer: the optimal price p is 7.65 when w = 0. The optimal weekday price is not within the credible range, since our range is between 1.5 and 7.5. Price(12.85 > max range 7.5). The optimal profit is 11.747.

### Part II: Pricing Analysis New vs Refurbished Product

In [52]:
df_refurb = pd.read_csv('refurb_data.csv')
df_refurb['dates'] = pd.to_datetime(df_refurb['dates'])
df_refurb['new_price'] = df_refurb['new_price'].astype("float")
df_refurb['refurb_price'] = df_refurb['refurb_price'].astype("float")
df_refurb.info()
df_refurb.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   dates         500 non-null    datetime64[ns]
 1   new_price     500 non-null    float64       
 2   refurb_price  500 non-null    float64       
 3   choice        500 non-null    object        
dtypes: datetime64[ns](1), float64(2), object(1)
memory usage: 15.8+ KB


dates  new_price  refurb_price  choice
0 2014-06-01      400.0         264.0  refurb
1 2014-06-01      400.0         264.0  refurb
2 2014-06-01      400.0         264.0  refurb
3 2014-06-01      400.0         264.0     new
4 2014-06-01      400.0         264.0     new
5 2014-06-01      400.0         264.0  refurb
6 2014-06-01      400.0         264.0  refurb
7 2014-06-01      400.0         264.0     new
8 2014-06-01      400.0         264.0     new
9 2014-06-01      400.0         264.0  refurb

In [88]:
df_re = df_refurb[df_refurb['choice'] == 'refurb']
df_re = df_re.groupby('dates').agg({'new_price': 'mean','refurb_price': 'mean', 'choice': 'count'}).reset_index()
df_re.columns = ['dates', 'new_price', 'refurb_price', 'refurb_count']

df_re.head(5)

dates  new_price  refurb_price  refurb_count
0 2014-06-01      400.0         264.0             9
1 2014-06-02      400.0         305.0             3
2 2014-06-03      400.0         306.0             6
3 2014-06-04      400.0         283.0             7
4 2014-06-05      400.0         305.0             8

In [90]:
import statsmodels.formula.api as smf
demand_reg = smf.ols(formula='refurb_count~ refurb_price + new_price ', data=df_re).fit()
demand_reg.summary()
demand_reg.params

Intercept      -0.701876
refurb_price   -0.044835
new_price       0.050579
dtype: float64

In [91]:
pi_300 = lambda x: -(demand_reg.params[0] + demand_reg.params[1]*x + demand_reg.params[2]*300 ) * (x-150)
from scipy.optimize import minimize_scalar
#optimal_price is x
minimize_scalar(pi_300, method='brent')

     fun: -334.6266140552393
    nfev: 11
     nit: 4
 success: True
       x: 236.39203136098254

In [95]:
pi_350 = lambda x: -(demand_reg.params[0] + demand_reg.params[1]*x + demand_reg.params[2]*350 ) * (x-150)
from scipy.optimize import minimize_scalar
#optimal_price is x
minimize_scalar(pi_350, method='brent')

     fun: -588.7710949351235
    nfev: 11
     nit: 4
 success: True
       x: 264.59526853520975

In [93]:
pi_400 = lambda x: -(demand_reg.params[0] + demand_reg.params[1]*x + demand_reg.params[2]*400 ) * (x-150)
from scipy.optimize import minimize_scalar
#optimal_price is x
minimize_scalar(pi_400, method='brent')

     fun: -914.2404513119158
    nfev: 11
     nit: 4
 success: True
       x: 292.7985057094416

### Answer: As shown above, the higher the new price, the higher the refurb price. When we have new price at 300, our refurb price is 236.392; When we have new price at 350, our refurb price is 264.595; When we have new price at 400, our refurb price is 292.799.
### For optimal profit, the higher the new price, the higher the optimal profit. When we have new price at 300, our max profit is 334.626; When we have new price at 350, our max profit is 588.77; When we have new price at 400, our max profit is 914.24.

## Part III: Pricing Analysis
## Two-Segment Pricing

First, find the optimal price* when only offering one price→i.e., with no segmentation

In [35]:
#Optimal price calcualtion with cost = $5
#max  = (10000 - 800p) *(p - 5)
#p(optimal price) = 8.75
# if offer one price
D = 10000 - 800 * 8.75
D
# Optimal profit is 11250

3000.0

### Answer:  when only offering one price the optimal price is $8.75

Report the optimal prices p* for segment 1 and segment 2

In [36]:
# Discussion with WTP < 7 & WTP > 7
#low segment size: 
Low_segment_size = (7/12.5)*10000
Low_segment_size
print("The low segment size is:", Low_segment_size)
High_segment_size = (12.5 - 7) / 12.5 *10000
print("The high segment size is:", High_segment_size)

The low segment size is: 5600.000000000001
The high segment size is: 4400.0


In [37]:
#Optimal price for the low demand
#low_segment = (5600 -800p)*(p-5) [Deriative this formula]
# price for low segement is $6
print('optimal price for low(WTP<7) segement is $6')
#high_segment = (4400 -800p)*(p-5) [Deriative this formula]
#We charge the optimal price when offering one uniqe price for high_segment
print('optimal price for high(WTP>7) segement is $8.75')


optimal price for low(WTP<7) segement is $6
optimal price for high(WTP>7) segement is $8.75


In [98]:
# Profit for segment #1
price = 8.75
demand = 3000
cost = 5
profit = price*demand - demand*cost
profit
print('The profit for high segment customer is ${:.2f}'.format(profit))
# Profit for segment #2
price_seg2 = 6
demand_seg2 = 800
cost = 5
profit_seg2 = price_seg2*demand_seg2 - demand_seg2*cost
profit_seg2
print('The profit for low segment customer is ${:.2f}'.format(profit1))
print('The profit for no segmentation case is $11250.00')

The profit for high segment customer is $11250.00
The profit for low segment customer is $800.00
The profit for no segmentation case is $11250.00


In [99]:
# Different in profit
diff = (profit + profit_seg2 - 11250) / profit *100
print('Different in profit in percentage is {:.2f}%'. format(diff))
print('Different in profit in percentage is $800')

Different in profit in percentage is 7.11%
Different in profit in percentage is $800


In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
sns.scatterplot(x = "price", y = "sales", data = df, color = "blue",hue="country")

In [ ]:
import statsmodels.formula.api as smf
import statsmodels.api as sm
country_dummy = pd.get_dummies(df["country"], prefix="country")
inventory_level_dummy = pd.get_dummies(df["inventory_level"], prefix="inventory_level")


# combine dummy variables and other variables
X = pd.concat([df[["price"]], country_dummy, inventory_level_dummy], axis=1)
X = sm.add_constant(X)

# fit the model and print summary
model = sm.OLS(df["sales"], X).fit()
print(model.summary())

In [ ]:
params.reset_index()

In [ ]:
#China at Optimal inventory level
from scipy.optimize import minimize_scalar
pi = lambda x: -(model.params[0] + model.params[1]*x + model.params[5]*1 + model.params[11]*1) * (x-0.6)
optimal_price_China = minimize_scalar(pi, method='brent')
print(optimal_price_China)

df_France = df.loc[df['country'] == 'France']

print(f'(Min) {df.price.min()}', ' < ', round(optimal_price_China.x,2), ' < ', f'{df.price.max()} (Max)')
df.price.min() < optimal_price_China.x < df.price.max()
optimal_p = 60 *2.149 - 60*0.6
optimal_p